### ISSUE TITLE
Self-referential Dict hangs Julia on display

### ISSUE THREAD

**Issue by mauro3** [JuliaLang/julia#10647]
```
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "help()" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.0-dev+4022 (2015-03-26 19:08 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 01a3216* (0 days old master)
|__/                   |  x86_64-unknown-linux-gnu

julia> oid = ObjectIdDict()
ObjectIdDict with 0 entries

julia> oid[oid] = 5
5

julia> oid
ObjectIdDict with 1 entry:Error showing value of type ObjectIdDict:
```
and just hangs, Ctrl-c has no effect.  In fact it leaves the julia process in a uninterruptible sleep state, i.e. I cannot even kill it:
```
  PID USER      PR  NI    VIRT    RES  %CPU %MEM     TIME+ S COMMAND                          
12875 mauro     20   0 8654.7m 243.2m   0.0  1.5   0:16.66 D julia0.4-pre                     
  593 mauro     20   0  358.1m 119.8m   0.0  0.8   0:02.05 D julia                            
```
(on Arch Linux)

Same for 0.3.6 and 0.2. Same for straight `Dict`.  Stumbled upon this when trying to see `Base.META`

Of course there is little sense in using a self-referential Dict.  But Julia should crash more gracefully.  Self-referential ObjectIdDicts should work though.

Related:
#3066 (test case in that issue works)
#7217, which now also causes an uninterruptible sleep state, and not a stack-overflow like back then.

**Comment by mauro3** [comment]
Update: after a computer update, restart & julia0.4 rebuild I do not get the hanging REPL anymore, instead I get a stackoverflow but the REPL stays usable.

**Comment by jakebolewski** [comment]
It is easy to print these without the stack overflow, we just have to adopt a standard way of doing it.  For types it is `#= circular reference =#` which is a bit verbose.  Python uses the `...` notation, perhaps we could use unicode ⟲?

**Comment by mbauman** [comment]
I was just going to mimic the generic types setup.  Personally, I like that it's printed in a block comment.

**Comment by StefanKarpinski** [comment]
I like ⟲ for this.

**Comment by amitmurthy** [comment]
+1 for  ⟲

**Comment by jiahao** [comment]
I guess the little space really makes ⟲ preferable to ⥀, eh. (#3831)

**Comment by StefanKarpinski** [comment]
This is an hilarious thread on my iPhone which doesn't have display versions of any these characters.

**Comment by jakebolewski** [comment]
closed by #10657.
